File: backend/services/prediction.py
import logging
import numpy as np
import yfinance as yf
import onnxruntime as rt
from sqlmodel import Session
from backend.db import engine
from backend.supabase import get_supabase
from backend.repositories import assets as assets_repository
from backend.repositories import asset_models as asset_models_repository
from backend.train.features import FEATURES, build_features
from backend.types.prediction import AssetPrediction

logger = logging.getLogger(__name__)

BUCKET = "ml-models"


def _download_model(storage_path: str) -> rt.InferenceSession:
    client = get_supabase()
    onnx_bytes = client.storage.from_(BUCKET).download(storage_path)
    return rt.InferenceSession(onnx_bytes)


def predict_asset(symbol: str) -> AssetPrediction:
    with Session(engine) as session:
        asset = assets_repository.get_asset_by_symbol(session, symbol)
        if asset is None:
            raise ValueError(f"Asset with symbol {symbol} not found")

        active_model = asset_models_repository.get_active_model(session, asset.id)
        if active_model is None:
            raise ValueError(f"No trained model found for {symbol}. Run /train first.")

        metrics = active_model.metrics
        storage_path = active_model.storage_path

    logger.info(f"Downloading model for {symbol} from {storage_path}")
    sess = _download_model(storage_path)

    ticker = yf.Ticker(asset.yfinance_symbol)
    df = ticker.history(period="60d")
    df = build_features(df)

    if df.empty:
        raise ValueError(f"Not enough data to build features for {symbol}")

    input_name = sess.get_inputs()[0].name
    output_name = sess.get_outputs()[1].name
    X = df[FEATURES].iloc[[-1]].values.astype(np.float32)
    proba = sess.run([output_name], {input_name: X})[0][0]

    signal = "BUY" if proba[1] >= 0.5 else "SELL"
    confidence = round(float(max(proba)), 4)

    return AssetPrediction(
        symbol=symbol,
        signal=signal,
        confidence=confidence,
        metrics=metrics,
    )

File: web/src/lib/services/prediction.ts
import YahooFinance from "yahoo-finance2";
import * as ort from "onnxruntime-node";
import { supabase } from "./supabase";
import { buildFeatures, FEATURES } from "../features";

const yf = new YahooFinance();
const BUCKET = "ml-models";

export type PredictionResult = {
  signal: "BUY" | "SELL";
  confidence: number;
  balanced_accuracy: number;
  roc_auc: number;
} | null;

async function getActiveModel(assetId: string) {
  const { data } = await supabase
    .from("asset_models")
    .select("storage_path, metrics")
    .eq("asset_id", assetId)
    .eq("is_active", true)
    .maybeSingle();
  return data;
}

// Returns null if no model exists — no training is triggered
export async function predictAsset(
  yfinanceSymbol: string,
  assetId: string,
): Promise<PredictionResult> {
  const model = await getActiveModel(assetId);
  console.log(
    `[predictAsset] assetId=${assetId} activeModel=`,
    model ? { path: model.storage_path, metrics: model.metrics } : "NONE",
  );
  if (!model) return null;

  const { data: blob, error } = await supabase.storage
    .from(BUCKET)
    .download(model.storage_path);
  console.log(
    `[predictAsset] download storage_path=${model.storage_path} error=`,
    error ?? "none",
    "blob size=",
    blob?.size ?? 0,
  );
  if (error || !blob) return null;

  const endDate = new Date();
  const startDate = new Date();
  startDate.setDate(startDate.getDate() - 90);

  let historical: {
    open: number;
    high: number;
    low: number;
    close: number;
    volume: number;
  }[] = [];
  try {
    const chart = await yf.chart(yfinanceSymbol, {
      period1: startDate,
      period2: endDate,
    });
    historical = chart.quotes.filter((h) => h.close !== null) as {
      open: number;
      high: number;
      low: number;
      close: number;
      volume: number;
    }[];
  } catch (e) {
    console.log(`[predictAsset] yfinance error:`, e);
    return null;
  }

  console.log(`[predictAsset] historical rows=${historical.length}`);
  if (historical.length < 30) return null;

  const records = historical.map((h) => ({
    open: h.open,
    high: h.high,
    low: h.low,
    close: h.close,
    volume: h.volume,
  }));
  const features = buildFeatures(records);
  if (features.length === 0) {
    console.log(`[predictAsset] feature engineering returned 0 rows`);
    return null;
  }

  const last = features[features.length - 1];
  const inputData = Float32Array.from(
    FEATURES.map((f) => last[f as keyof typeof last] as number),
  );

  try {
    const modelBytes = new Uint8Array(await blob.arrayBuffer());
    console.log(`[predictAsset] ONNX model bytes=${modelBytes.length}`);
    const session = await ort.InferenceSession.create(modelBytes);
    console.log(
      `[predictAsset] ONNX session created. inputs=${session.inputNames} outputs=${session.outputNames}`,
    );
    const inputTensor = new ort.Tensor("float32", inputData, [
      1,
      FEATURES.length,
    ]);
    const results = await session.run({ [session.inputNames[0]]: inputTensor });
    const proba = results[session.outputNames[1]].data as Float32Array;
    console.log(`[predictAsset] proba=`, Array.from(proba));

    const signal = proba[1] >= 0.5 ? "BUY" : "SELL";
    const confidence = Math.round(Math.max(proba[0], proba[1]) * 10000) / 10000;

    return {
      signal,
      confidence,
      balanced_accuracy:
        (model.metrics as Record<string, number>)?.balanced_accuracy ?? 0,
      roc_auc: (model.metrics as Record<string, number>)?.roc_auc ?? 0,
    };
  } catch (e) {
    console.log(`[predictAsset] ONNX inference error:`, e);
    return null;
  }
}
